invalid suggestion to add `.into_iter()` into the source code of `quote`

### Code

```Rust
use proc_macro::TokenStream;

#[proc_macro]
pub fn default_args_fn(_: TokenStream) -> TokenStream {
    let decl_macro = TokenStream::new();

    quote::quote! {
        #(#decl_macro)*
    }
    .into()
}
```

### Current output

```Shell
error[E0599]: no method named `quote_into_iter` found for struct `proc_macro::TokenStream` in the current scope
 --> src/lib.rs:7:5
  |
7 | /     quote::quote! {
8 | |         #(#decl_macro)*
9 | |     }
  | |_____^ `proc_macro::TokenStream` is not an iterator
  |
  = note: this error originates in the macro `$crate::quote_bind_into_iter` which comes from the expansion of the macro `quote::quote` (in Nightly builds, run with -Z macro-backtrace for more info)
help: call `.into_iter()` first
 --> /home/user/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/quote-1.0.38/src/lib.rs:744:34
  |
74|         let (mut $var, i) = $var.into_iter().quote_into_iter();
  |                                  ++++++++++++

For more information about this error, try `rustc --explain E0599`.
```

### Desired output

```Shell
editing foreign code should not be suggested
```

### Rationale and extra context

_No response_

### Other cases

```Rust

```

### Rust Version

```Shell
rustc 1.87.0-nightly (f280acf4c 2025-02-19)
binary: rustc
commit-hash: f280acf4c743806abbbbcfe65050ac52ec4bdec0
commit-date: 2025-02-19
host: x86_64-unknown-linux-gnu
release: 1.87.0-nightly
LLVM version: 20.1.0
```

### Anything else?

_No response_

--- a/compiler/rustc_hir_typeck/src/method/suggest.rs
+++ b/compiler/rustc_hir_typeck/src/method/suggest.rs
@@ -905,11 +905,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         } else if self.impl_into_iterator_should_be_iterator(rcvr_ty, span, unsatisfied_predicates)
         {
             err.span_label(span, format!("`{rcvr_ty}` is not an iterator"));
-            err.multipart_suggestion_verbose(
-                "call `.into_iter()` first",
-                vec![(span.shrink_to_lo(), format!("into_iter()."))],
-                Applicability::MaybeIncorrect,
-            );
+            if !span.in_external_macro(self.tcx.sess.source_map()) {
+                err.multipart_suggestion_verbose(
+                    "call `.into_iter()` first",
+                    vec![(span.shrink_to_lo(), format!("into_iter()."))],
+                    Applicability::MaybeIncorrect,
+                );
+            }
             return err.emit();
         } else if !unsatisfied_predicates.is_empty() && matches!(rcvr_ty.kind(), ty::Param(_)) {
             // We special case the situation where we are looking for `_` in

--- a/tests/ui/proc-macro/auxiliary/quote-issue-137345.rs
+++ b/tests/ui/proc-macro/auxiliary/quote-issue-137345.rs
@@ -1,0 +1,19 @@
+extern crate proc_macro;
+use proc_macro::TokenStream;
+
+fn items() -> impl IntoIterator<Item = i32> {
+    vec![1, 2, 3]
+}
+
+#[macro_export]
+macro_rules! quote {
+    // Rule for any other number of tokens.
+    ($($tt:tt)*) => {{
+        fn items() -> impl IntoIterator<Item = i32> {
+            vec![1, 2, 3]
+        }
+        let _s = TokenStream::new();
+        let other_items = items().map(|i| i + 1);
+        _s
+    }};
+}

--- a/tests/ui/proc-macro/valid-sugg-issue-137345.rs
+++ b/tests/ui/proc-macro/valid-sugg-issue-137345.rs
@@ -1,0 +1,17 @@
+//@ aux-crate: quote=quote-issue-137345.rs
+
+extern crate proc_macro;
+extern crate quote;
+
+use proc_macro::TokenStream;
+
+pub fn default_args_fn(_: TokenStream) -> TokenStream {
+    let decl_macro = TokenStream::new();
+
+    quote::quote! {
+        #(#decl_macro)*
+    }
+    .into() //~^^^ ERROR no method named `map` found for opaque type
+}
+
+fn main() {}

--- a/tests/ui/proc-macro/valid-sugg-issue-137345.stderr
+++ b/tests/ui/proc-macro/valid-sugg-issue-137345.stderr
@@ -1,0 +1,13 @@
+error[E0599]: no method named `map` found for opaque type `impl IntoIterator<Item = i32>` in the current scope
+  --> $DIR/valid-sugg-issue-137345.rs:11:5
+   |
+LL | /     quote::quote! {
+LL | |         #(#decl_macro)*
+LL | |     }
+   | |_____^ `impl IntoIterator<Item = i32>` is not an iterator
+   |
+   = note: this error originates in the macro `quote::quote` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0599`.

